---
title: "RNAseq Single Cell Clustering"
author: "David Schulz"
date: "7/3/2018"
output:
  pdf_document: default
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(pvclust)
library(heatmap.plus)
```



```{r pvclust Allcontigs1}

counts <- read.csv("singleCell_kallisto_counts.csv", row.names = "id")
countsALL28k <- subset(counts, rowSums(counts[,1:37])>0 )
countsALL28k <- t(countsALL28k)
countsALL28k <- scale(countsALL28k)
countsALL28k <- t(countsALL28k)

contigALL28kpvclust1 <- pvclust(countsALL28k, nboot = 10, use.cor="pairwise.complete.obs", method.dist = "correlation", method.hclust = "ward.D")
contigALL28kpvclust2 <- pvclust(countsALL28k, nboot = 10, use.cor="pairwise.complete.obs", method.dist = "manhattan", method.hclust = "ward.D")
contigALL28kpvclust3 <- pvclust(countsALL28k, nboot = 10, use.cor="pairwise.complete.obs", method.dist = "euclidean", method.hclust = "ward.D")


plot(contigALL28kpvclust1, main = "28kcontigs scaled by contig")
plot(contigALL28kpvclust2, main = "28kcontigs scaled by contig")
plot(contigALL28kpvclust3, main = "28kcontigs scaled by contig")

#pvrect(contigALL28kpvclust, alpha=0.95)

#heatmap.plus::heatmap.plus(as.matrix(countsALL28k), hclustfun = function(x) hclust(x,method = 'ward.D'), distfun = function(y) dist(y,method = 'manhattan'))
```

```{r pvclust Allcontigs2}

counts <- read.csv("singleCell_kallisto_counts.csv", row.names = "id")
countsALL28k <- subset(counts, rowSums(counts[,1:37])>0 )
#countsALL28k <- t(countsALL28k)
countsALL28k <- scale(countsALL28k)
#countsALL28k <- t(countsALL28k)

contigALL28kpvclust <- pvclust(countsALL28k, nboot = 10, use.cor="pairwise.complete.obs", method.dist = "correlation", method.hclust = "ward.D")
plot(contigALL28kpvclust, main = "28kcontigs scaled by cell")


#pvrect(contigALL28kpvclust, alpha=0.95)

#heatmap.plus::heatmap.plus(as.matrix(countsALL28k), hclustfun = function(x) hclust(x,method = 'ward.D'), distfun = function(y) dist(y,method = 'manhattan'))
```


```{r pvclust 0.2contigs1}

counts <- read.csv("singleCell_kallisto_counts.csv", row.names = "id")
names <- read.csv("annotated_pooled_0.2_names.csv")
names <- t(names)
x <- counts[c(names),]
 

counts <- subset(x, rowSums(counts[,1:37])>0 )
#counts <- na.omit(counts)
counts <- t(counts)
counts <- scale(counts)
counts <- t(counts)

contig0.2pvclust1 <- pvclust(counts, nboot = 10, use.cor="pairwise.complete.obs", method.dist = "correlation", method.hclust = "ward.D")
contig0.2pvclust2 <- pvclust(counts, nboot = 10, use.cor="pairwise.complete.obs", method.dist = "manhattan", method.hclust = "ward.D")
contig0.2pvclust3 <- pvclust(counts, nboot = 10, use.cor="pairwise.complete.obs", method.dist = "euclidean", method.hclust = "ward.D")


plot(contig0.2pvclust1, main = "0.2contigs scaled by contig")
plot(contig0.2pvclust2, main = "0.2contigs scaled by contig")
plot(contig0.2pvclust3, main = "0.2contigs scaled by contig")

#pvrect(contig0.2pvclust, alpha=0.95)
#heatmap.plus::heatmap.plus(as.matrix(counts), hclustfun = function(x) hclust(x,method = 'ward.D'), distfun = function(y) dist(y,method = 'manhattan'))
```

```{r pvclust 0.2contigs2}

counts <- read.csv("singleCell_kallisto_counts.csv", row.names = "id")
names <- read.csv("annotated_pooled_0.2_names.csv")
names <- t(names)
x <- counts[c(names),]
 

counts <- subset(x, rowSums(counts[,1:37])>0 )
#counts <- na.omit(counts)
#counts <- t(counts)
counts <- scale(counts)
#counts <- t(counts)

contig0.2pvclust1 <- pvclust(counts, nboot = 10, use.cor="pairwise.complete.obs", method.dist = "correlation", method.hclust = "ward.D")
contig0.2pvclust2 <- pvclust(counts, nboot = 10, use.cor="pairwise.complete.obs", method.dist = "manhattan", method.hclust = "ward.D")
contig0.2pvclust3 <- pvclust(counts, nboot = 10, use.cor="pairwise.complete.obs", method.dist = "euclidean", method.hclust = "ward.D")


plot(contig0.2pvclust1, main = "0.2contigs scaled by cell")
plot(contig0.2pvclust2, main = "0.2contigs scaled by cell")
plot(contig0.2pvclust3, main = "0.2contigs scaled by cell")

#pvrect(contig0.2pvclust, alpha=0.95)
#heatmap.plus::heatmap.plus(as.matrix(counts), hclustfun = function(x) hclust(x,method = 'ward.D'), distfun = function(y) dist(y,method = 'manhattan'))
```


```{r pvclust 0.05contigs1}

counts <- read.csv("singleCell_kallisto_counts.csv", row.names = "id")
names <- read.csv("annotated_pooled_0.05_names.csv")
names <- t(names)
x <- counts[c(names),]
cells <- read.csv("sloober.csv") 

counts <- subset(x, rowSums(counts[,1:37])>0 )
#counts <- na.omit(counts)
counts <- t(counts)
counts <- scale(counts)
counts <- t(counts)

contig0.05pvclust1 <- pvclust(counts, nboot = 10, use.cor="pairwise.complete.obs", method.dist = "correlation", method.hclust = "ward.D")
contig0.05pvclust2 <- pvclust(counts, nboot = 10, use.cor="pairwise.complete.obs", method.dist = "manhattan", method.hclust = "ward.D")
contig0.05pvclust3 <- pvclust(counts, nboot = 10, use.cor="pairwise.complete.obs", method.dist = "euclidean", method.hclust = "ward.D")


plot(contig0.05pvclust1, main = "0.05contigs scaled by contig")
plot(contig0.05pvclust2, main = "0.05contigs scaled by contig")
plot(contig0.05pvclust3, main = "0.05contigs scaled by contig")



#pvrect(contig0.05pvclust, alpha=0.95)

#heatmap.plus::heatmap.plus(as.matrix(counts), hclustfun = function(x) hclust(x,method = 'ward.D'), distfun = function(y) dist(y,method = 'manhattan'))
```
```{r pvclust 0.05contigs2}

counts <- read.csv("singleCell_kallisto_counts.csv", row.names = "id")
names <- read.csv("annotated_pooled_0.05_names.csv")
names <- t(names)
x <- counts[c(names),]
cells <- read.csv("sloober.csv") 

counts <- subset(x, rowSums(counts[,1:37])>0 )
#counts <- na.omit(counts)
#counts <- t(counts)
counts <- scale(counts)
#counts <- t(counts)

contig0.05pvclust1 <- pvclust(counts, nboot = 10, use.cor="pairwise.complete.obs", method.dist = "correlation", method.hclust = "ward.D")
contig0.05pvclust2 <- pvclust(counts, nboot = 10, use.cor="pairwise.complete.obs", method.dist = "manhattan", method.hclust = "ward.D")
contig0.05pvclust3 <- pvclust(counts, nboot = 10, use.cor="pairwise.complete.obs", method.dist = "euclidean", method.hclust = "ward.D")


plot(contig0.05pvclust1, main = "0.05contigs scaled by cell")
plot(contig0.05pvclust2, main = "0.05contigs scaled by cell")
plot(contig0.05pvclust3, main = "0.05contigs scaled by cell")



#pvrect(contig0.05pvclust, alpha=0.95)

#heatmap.plus::heatmap.plus(as.matrix(counts), hclustfun = function(x) hclust(x,method = 'ward.D'), distfun = function(y) dist(y,method = 'manhattan'))
```